modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionDemander.java

```diff
@@ -1309,9 +1335,29 @@ private void updateGroupMetrics() {
             this.rebalanceId = -1;
             this.routines = 0;
             this.cancelLock = new ReentrantReadWriteLock();
+            this.next = null;
             this.lastCancelledTime = new AtomicLong();
         }
 
+        /**
+         * Init future.
+         */
+        public void init() {
+            requestPartitions(this, assignments);
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean onDone(@Nullable Boolean res, @Nullable Throwable err) {
+            if (super.onDone(res, err)) {
+                if (next != null)
```

Comment:
Is it true that the `next` will start regardless the current RebalanceFuture finishes with an error or cancelled e.g. on stop() method called? Is it correct?

Reply:
I think yes. All rebalance's futures are executed independently of other.

Reply:
It is OK that rebalance futures are independent, but it's strange when the current future has cancelled, but the next cache group starts nevertheless. We should:
1. cancel the whole chain since we are building one 
2. use an external arbiter (an executor service) for such tasks.

Please also note that parallel rebalancing of cache groups will flood the whole cluster an can lead to the cluster death.